File: lunar_calendar.go
// lunar_calendar.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

var zodiacSigns = []string{"Овен", "Телец", "Близнецы", "Рак", "Лев", "Дева",
	"Весы", "Скорпион", "Стрелец", "Козерог", "Водолей", "Рыбы"}

var recommendations = map[string]map[string]string{
	"новолуние":          {"посадка": "неблагоприятно", "полив": "допустимо", "обрезка": "неблагоприятно", "урожай": "неблагоприятно", "вредители": "благоприятно"},
	"первая четверть":    {"посадка": "благоприятно", "полив": "благоприятно", "обрезка": "неблагоприятно", "урожай": "неблагоприятно", "вредители": "неблагоприятно"},
	"полнолуние":         {"посадка": "неблагоприятно", "полив": "допустимо", "обрезка": "благоприятно", "урожай": "благоприятно", "вредители": "благоприятно"},
	"последняя четверть": {"посадка": "неблагоприятно", "полив": "неблагоприятно", "обрезка": "благоприятно", "урожай": "благоприятно", "вредители": "благоприятно"},
}

type LunarData struct {
	Date          string            `json:"date"`
	LunarDay      int               `json:"lunar_day"`
	Phase         string            `json:"phase"`
	ZodiacSign    string            `json:"zodiac_sign"`
	Recommendations map[string]string `json:"recommendations"`
}

func parseDate(dateStr string) time.Time {
	if dateStr == "" {
		return time.Now()
	}
	t, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Ошибка парсинга даты: %v\n", err)
		os.Exit(1)
	}
	return t
}

func calculate(date time.Time) LunarData {
	// Базовое новолуние: 2000-01-06 18:14 UTC
	base := time.Date(2000, 1, 6, 18, 14, 0, 0, time.UTC)
	diff := date.Sub(base).Hours() / 24
	lunarAge := diff - math.Floor(diff/29.53058867)*29.53058867
	lunarDay := int(lunarAge) + 1
	if lunarDay > 30 {
		lunarDay = 30
	}
	var phase string
	if lunarDay <= 1 || lunarDay > 29 {
		phase = "новолуние"
	} else if lunarDay <= 7 {
		phase = "первая четверть"
	} else if lunarDay <= 14 {
		phase = "полнолуние"
	} else if lunarDay <= 22 {
		phase = "последняя четверть"
	} else {
		phase = "новолуние"
	}
	// Знак зодиака
	start := time.Date(2000, 1, 1, 0, 0, 0, 0, time.UTC)
	totalDays := date.Sub(start).Hours() / 24
	longitude := 180 + totalDays*13.176
	longitude = math.Mod(longitude, 360)
	signIndex := int(math.Floor(longitude / 30)) % 12
	sign := zodiacSigns[signIndex]
	return LunarData{
		Date:          date.Format("2006-01-02"),
		LunarDay:      lunarDay,
		Phase:         phase,
		ZodiacSign:    sign,
		Recommendations: recommendations[phase],
	}
}

func printData(data LunarData, verbose bool, color bool) {
	if color {
		fmt.Printf("\033[36m🌙 Лунный календарь на %s\033[0m\n", data.Date)
		fmt.Printf("\033[33mЛунный день: %d\033[0m\n", data.LunarDay)
		fmt.Printf("\033[35mФаза: %s\033[0m\n", data.Phase)
		fmt.Printf("\033[32mЗнак зодиака Луны: %s\033[0m\n", data.ZodiacSign)
		if verbose {
			fmt.Println("\033[37mРекомендации:\033[0m")
			for k, v := range data.Recommendations {
				var col string
				if v == "благоприятно" {
					col = "\033[32m"
				} else if v == "неблагоприятно" {
					col = "\033[31m"
				} else {
					col = "\033[33m"
				}
				fmt.Printf("  - %s: %s%s\033[0m\n", k, col, v)
			}
		}
	} else {
		fmt.Printf("🌙 Лунный календарь на %s\n", data.Date)
		fmt.Printf("Лунный день: %d\n", data.LunarDay)
		fmt.Printf("Фаза: %s\n", data.Phase)
		fmt.Printf("Знак зодиака Луны: %s\n", data.ZodiacSign)
		if verbose {
			fmt.Println("Рекомендации:")
			for k, v := range data.Recommendations {
				fmt.Printf("  - %s: %s\n", k, v)
			}
		}
	}
}

func main() {
	var (
		dateStr string
		jsonOut string
		csvOut  string
		verbose bool
		noColor bool
	)
	flag.StringVar(&dateStr, "date", "", "Дата (YYYY-MM-DD)")
	flag.StringVar(&jsonOut, "json", "", "Экспорт в JSON")
	flag.StringVar(&csvOut, "csv", "", "Экспорт в CSV")
	flag.BoolVar(&verbose, "verbose", false, "Подробный вывод")
	flag.BoolVar(&noColor, "no-color", false, "Отключить цвет")
	flag.Parse()

	date := parseDate(dateStr)
	data := calculate(date)
	color := !noColor && isTerminal()
	printData(data, verbose, color)

	if jsonOut != "" {
		jsonData, err := json.MarshalIndent(data, "", "  ")
		if err != nil {
			fmt.Fprintf(os.Stderr, "Ошибка JSON: %v\n", err)
		} else {
			os.WriteFile(jsonOut, jsonData, 0644)
			fmt.Printf("Результат сохранён в %s\n", jsonOut)
		}
	}

	if csvOut != "" {
		f, err := os.Create(csvOut)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Ошибка CSV: %v\n", err)
		} else {
			defer f.Close()
			w := csv.NewWriter(f)
			defer w.Flush()
			w.Write([]string{"date", "lunar_day", "phase", "zodiac_sign", "recommendations"})
			recStr := ""
			for k, v := range data.Recommendations {
				if recStr != "" {
					recStr += "; "
				}
				recStr += k + ":" + v
			}
			w.Write([]string{data.Date, strconv.Itoa(data.LunarDay), data.Phase, data.ZodiacSign, recStr})
			fmt.Printf("Результат сохранён в %s\n", csvOut)
		}
	}
}

func isTerminal() bool {
	stat, _ := os.Stdout.Stat()
	return (stat.Mode() & os.ModeCharDevice) != 0
}

File: lunar_calendar.rs
// lunar_calendar.rs
use chrono::{DateTime, NaiveDate, Utc, TimeZone};
use clap::{App, Arg};
use serde::{Serialize, Deserialize};
use serde_json;
use std::fs;
use std::io::Write;
use std::collections::HashMap;
use colored::*;

const ZODIAC_SIGNS: [&str; 12] = ["Овен", "Телец", "Близнецы", "Рак", "Лев", "Дева",
                                  "Весы", "Скорпион", "Стрелец", "Козерог", "Водолей", "Рыбы"];

fn recommendations() -> HashMap<String, HashMap<String, String>> {
    let mut map = HashMap::new();
    map.insert("новолуние".to_string(), {
        let mut m = HashMap::new();
        m.insert("посадка".to_string(), "неблагоприятно".to_string());
        m.insert("полив".to_string(), "допустимо".to_string());
        m.insert("обрезка".to_string(), "неблагоприятно".to_string());
        m.insert("урожай".to_string(), "неблагоприятно".to_string());
        m.insert("вредители".to_string(), "благоприятно".to_string());
        m
    });
    map.insert("первая четверть".to_string(), {
        let mut m = HashMap::new();
        m.insert("посадка".to_string(), "благоприятно".to_string());
        m.insert("полив".to_string(), "благоприятно".to_string());
        m.insert("обрезка".to_string(), "неблагоприятно".to_string());
        m.insert("урожай".to_string(), "неблагоприятно".to_string());
        m.insert("вредители".to_string(), "неблагоприятно".to_string());
        m
    });
    map.insert("полнолуние".to_string(), {
        let mut m = HashMap::new();
        m.insert("посадка".to_string(), "неблагоприятно".to_string());
        m.insert("полив".to_string(), "допустимо".to_string());
        m.insert("обрезка".to_string(), "благоприятно".to_string());
        m.insert("урожай".to_string(), "благоприятно".to_string());
        m.insert("вредители".to_string(), "благоприятно".to_string());
        m
    });
    map.insert("последняя четверть".to_string(), {
        let mut m = HashMap::new();
        m.insert("посадка".to_string(), "неблагоприятно".to_string());
        m.insert("полив".to_string(), "неблагоприятно".to_string());
        m.insert("обрезка".to_string(), "благоприятно".to_string());
        m.insert("урожай".to_string(), "благоприятно".to_string());
        m.insert("вредители".to_string(), "благоприятно".to_string());
        m
    });
    map
}

#[derive(Serialize, Deserialize)]
struct LunarData {
    date: String,
    lunar_day: i32,
    phase: String,
    zodiac_sign: String,
    recommendations: HashMap<String, String>,
}

fn calculate(date: NaiveDate) -> LunarData {
    // Базовое новолуние: 2000-01-06 18:14 UTC
    let base = Utc.with_ymd_and_hms(2000, 1, 6, 18, 14, 0).unwrap();
    let dt = Utc.from_utc_datetime(&date.and_hms_opt(0,0,0).unwrap());
    let diff = dt - base;
    let days = diff.num_seconds() as f64 / 86400.0;
    let lunar_age = days % 29.53058867;
    let lunar_day = (lunar_age.floor() as i32) + 1;
    let lunar_day = if lunar_day > 30 { 30 } else { lunar_day };
    let phase = if lunar_day <= 1 || lunar_day > 29 {
        "новолуние".to_string()
    } else if lunar_day <= 7 {
        "первая четверть".to_string()
    } else if lunar_day <= 14 {
        "полнолуние".to_string()
    } else if lunar_day <= 22 {
        "последняя четверть".to_string()
    } else {
        "новолуние".to_string()
    };
    // Знак зодиака
    let start = Utc.with_ymd_and_hms(2000, 1, 1, 0, 0, 0).unwrap();
    let total_days = (dt - start).num_seconds() as f64 / 86400.0;
    let longitude = (180.0 + total_days * 13.176) % 360.0;
    let sign_index = ((longitude / 30.0).floor() as usize) % 12;
    let zodiac_sign = ZODIAC_SIGNS[sign_index].to_string();

    let rec_map = recommendations();
    let recs = rec_map.get(&phase).cloned().unwrap_or(HashMap::new());

    LunarData {
        date: date.format("%Y-%m-%d").to_string(),
        lunar_day,
        phase,
        zodiac_sign,
        recommendations: recs,
    }
}

fn print_data(data: &LunarData, verbose: bool, color: bool) {
    if color {
        println!("{}", format!("🌙 Лунный календарь на {}", data.date).cyan());
        println!("{}", format!("Лунный день: {}", data.lunar_day).yellow());
        println!("{}", format!("Фаза: {}", data.phase).magenta());
        println!("{}", format!("Знак зодиака Луны: {}", data.zodiac_sign).green());
        if verbose {
            println!("{}", "Рекомендации:".white());
            for (k, v) in &data.recommendations {
                let col = if v == "благоприятно" { "green" } else if v == "неблагоприятно" { "red" } else { "yellow" };
                let colored = match col {
                    "green" => v.green(),
                    "red" => v.red(),
                    _ => v.yellow(),
                };
                println!("  - {}: {}", k, colored);
            }
        }
    } else {
        println!("🌙 Лунный календарь на {}", data.date);
        println!("Лунный день: {}", data.lunar_day);
        println!("Фаза: {}", data.phase);
        println!("Знак зодиака Луны: {}", data.zodiac_sign);
        if verbose {
            println!("Рекомендации:");
            for (k, v) in &data.recommendations {
                println!("  - {}: {}", k, v);
            }
        }
    }
}

fn main() {
    let matches = App::new("Лунный календарь")
        .arg(Arg::with_name("date").long("date").takes_value(true).help("Дата YYYY-MM-DD"))
        .arg(Arg::with_name("json").long("json").takes_value(true).help("Экспорт в JSON"))
        .arg(Arg::with_name("csv").long("csv").takes_value(true).help("Экспорт в CSV"))
        .arg(Arg::with_name("verbose").long("verbose").help("Подробный вывод"))
        .arg(Arg::with_name("no-color").long("no-color").help("Отключить цвет"))
        .get_matches();

    let date_str = matches.value_of("date").unwrap_or("");
    let date = if date_str.is_empty() {
        chrono::Local::now().date_naive()
    } else {
        NaiveDate::parse_from_str(date_str, "%Y-%m-%d").expect("Ошибка формата даты")
    };
    let data = calculate(date);
    let color = !matches.is_present("no-color") && atty::is(atty::Stream::Stdout);
    print_data(&data, matches.is_present("verbose"), color);

    if let Some(json_file) = matches.value_of("json") {
        let json = serde_json::to_string_pretty(&data).unwrap();
        fs::write(json_file, json).unwrap();
        println!("Результат сохранён в {}", json_file);
    }

    if let Some(csv_file) = matches.value_of("csv") {
        let mut wtr = csv::Writer::from_path(csv_file).unwrap();
        wtr.write_record(&["date", "lunar_day", "phase", "zodiac_sign", "recommendations"]).unwrap();
        let rec_str = data.recommendations.iter()
            .map(|(k,v)| format!("{}:{}", k, v))
            .collect::<Vec<_>>().join("; ");
        wtr.write_record(&[data.date, data.lunar_day.to_string(), data.phase, data.zodiac_sign, rec_str]).unwrap();
        wtr.flush().unwrap();
        println!("Результат сохранён в {}", csv_file);
    }
}
